File: 19.cpp
//
// Created by linkinpark213 on 5/6/19.
//

#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;

    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        int count = 0;
        ListNode *ptr = head;
        while (ptr != NULL) {
            count++;
            ptr = ptr->next;
        }
        int target = count - n;
        ptr = head;
        for (int i = 0; i < target - 1; i++) {
            ptr = ptr->next;
        }
        if (target == 0) {
            return ptr->next;
        }
        ListNode *toRemove = ptr->next;
        ListNode *cont;
        if (toRemove != NULL)
            cont = toRemove->next;
        else
            cont = NULL;
        ptr->next = cont;
        return head;
    }
};

int main() {
    Solution solution;
    ListNode *head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = new ListNode(5);
    ListNode *ptr = solution.removeNthFromEnd(head, 2);
    while (ptr != NULL) {
        cout << ptr->val << ' ';
        ptr = ptr->next;
    }
}

File: pysrc/19.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

    def __str__(self):
        ptr = self
        s = ''
        while ptr is not None:
            s += str(ptr.val) + ', '
            ptr = ptr.next
        return s

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        ptr1 = dummy
        ptr2 = dummy
        for i in range(n + 1):
            ptr1 = ptr1.next
        while ptr1 is not None:
            ptr1 = ptr1.next
            ptr2 = ptr2.next

        ptr2.next = ptr2.next.next
        return dummy.next


if __name__ == '__main__':
    solution = Solution()
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    print(solution.removeNthFromEnd(head, 2))
